Assignment07/step-by-step-python.md

```diff
@@ -0,0 +1,217 @@
+
+The `nestedSeparator` in the `metadata` section specifies the character that Dapr will use when it flattens the secret  
+file's hierarchy. Eventually, each secret will be uniquely identifiable by one key. In this case, you're using the
+period (`.`) as character. That means that the secrets from the `secrets.json` file will be identified by the following
+keys:
+
+- `smtp.user`
+- `smtp.password`
+- `finecalculator.licensekey`
+
+Now you've configured the secrets management building block. Time to use the secrets.
+
+## Step 2: Get the credentials for connecting to the SMTP server
+
+As stated, you can reference secrets from other Dapr component configuration files.
+
+1. Open the file `python/dapr/components/email.yaml` in VS Code.
+
+1. Inspect the contents of the file. As you can see, it contains clear-text credentials (username and password).
+   Replace the `user` and `password` fields of the `metadata` with secret references and add a reference to the secrets
+   management building block named `trafficcontrol-secrets` you configured in step 1. The resulting file should look
+   like this:
+
+   ```yaml
+   apiVersion: dapr.io/v1alpha1
+   kind: Component
+   metadata:
+     name: sendmail
+   spec:
+     type: bindings.smtp
+     version: v1
+     metadata:
+     - name: host
+       value: localhost
+     - name: port
+       value: 4025
+     - name: user
+       secretKeyRef:
+         name: smtp.user
+         key: smtp.user
+     - name: password
+       secretKeyRef:
+         name: smtp.password
+         key: smtp.password
+     - name: skipTLSVerify
+       value: true
+   auth:
+     secretStore: trafficcontrol-secrets
+   scopes:
+     - finecollectionservice  
+   ```
+
+Now, the output binding will use the `smtp.username` and `smtp.password` secrets from the secrets file at runtime.
+
+## Step 3: Get the license key for the FineCalculator component
+
+The `ViolationProcessor` of the FineCollectionService uses a `FineCalculator` implementation to calculate the fine for
+a certain speeding violation (check out the code). The calculator used is implemented in the
+`python\FineCollectionService\fine_collection\services.py` file. It checks a license key (because we can). This license
+key has to be loaded from a secret since we don't want anyone stealing the key.
+
+Currently we load the license key from `python/FineCollectionService/.env` which isn't secure. We're going to change
+the application so it loads the license key as a secret.
+
+1. Open the file `python\FineCollectionService\fine_collection\__init__.py` in VS Code.
+
+2. Add the following import to the top of the file:
+
+   ```python
+   from dapr.clients import DaprClient
+   ```
+
+3. Replace the assignment of the `license_key` variable with the following code:
+
+```python
+with DaprClient() as dapr_client:
+    license_key = dapr_client.get_secret("trafficcontrol-secrets", "finecalculator.licensekey")
```

Comment:
I think this needs some attention. When I followed the steps, processing a speeding violation would result in the following snacktrace:

```console
== APP ==   File "/home/maarten/Code/dapr-workshop-python/python/FineCollectionService/./fine_collection/__init__.py", line 32, in collect_fine
== APP ==     processor.process_speed_violation(violation)
== APP ==   File "/home/maarten/Code/dapr-workshop-python/python/FineCollectionService/./fine_collection/services.py", line 43, in process_speed_violation
== APP ==     fine = self.calculator.calculate_fine(violation.violationInKmh)
== APP ==   File "/home/maarten/Code/dapr-workshop-python/python/FineCollectionService/./fine_collection/services.py", line 13, in calculate_fine
== APP ==     raise ValueError("Invalid license key")
== APP == ValueError: Invalid license key
```

To debug, I added `print("Initialising with license key " + license_key)` to the `__init__` method of the `FineCalculator`, but that prevents the Fine Collection Service from booting:

```console
== APP ==   File "/home/maarten/Code/dapr-workshop-python/python/FineCollectionService/./fine_collection/__init__.py", line 16, in <module>
== APP ==     services.FineCalculator(license_key),
== APP ==   File "/home/maarten/Code/dapr-workshop-python/python/FineCollectionService/./fine_collection/services.py", line 7, in __init__
== APP ==     print("Initialising with license key " + license_key)
== APP == TypeError: can only concatenate str (not "GetSecretResponse") to str
```

I think the `dapr_client.get_secret` method does not return a `str` representation of the secret, but rather some data transfer object that needs to be unpacked.

Reply:
Resolved with b63c337